framework/src/variables/MooseVariableDataBase.C

```diff
@@ -620,66 +606,49 @@ MooseVariableDataBase<OutputType>::fetchDoFValues()
 
 template <typename OutputType>
 void
-MooseVariableDataBase<OutputType>::getArrayDoFValues(const NumericVector<Number> & sol,
-                                                     unsigned int n,
+MooseVariableDataBase<OutputType>::getArrayDofValues(const NumericVector<Number> & sol,
+                                                     const unsigned int nshapes,
                                                      MooseArray<RealEigenVector> & dof_values) const
 {
-  dof_values.resize(n);
-  if (isNodal())
+  dof_values.resize(nshapes);
+  for (const auto i : make_range(nshapes))
   {
-    for (unsigned int i = 0; i < n; ++i)
-    {
-      dof_values[i].resize(_count);
-      auto dof = _dof_indices[i];
-      for (unsigned int j = 0; j < _count; ++j)
-        dof_values[i](j) = sol(dof++);
-    }
-  }
-  else
-  {
-    for (unsigned int i = 0; i < n; ++i)
-    {
-      dof_values[i].resize(_count);
-      auto dof = _dof_indices[i];
-      for (unsigned int j = 0; j < _count; ++j)
-      {
-        dof_values[i](j) = sol(dof);
-        dof += n;
-      }
-    }
+    dof_values[i].resize(_count);
+    for (const auto j : make_range(_count))
+      dof_values[i](j) = sol(_dof_indices[j * nshapes + i]);
```

Comment:
so why is there no concern that the vector be an array of lagrange_vec and that for each shape we need to loop from 0 to _count * 3 in this routine?

Reply:
we are using the `_dof_indices`. We're directly indexing into `_dof_indices` here. We mandate in libMesh that the array dof indexing has major ordering based on count and minor ordering based on shape

Reply:
We pick this convention because, for example, users can specify different scaling factors for different array variable components. It is easy to extract contiguous submatrices and subvectors corresponding to a given array variable component using the convention we've chosen for multiplication by these potentially disparate scaling factors